Support for class-style component as persistent layout

TheDoctor0 opened the pull request:
This PR will add support for class-style component (written in TypeScript) used as persistent layout.

Class based VueComponent have different structure and component object is not available 
on `default` level, as it it actually stored in nested `options` property.
It causes an error in Inertia render function: `TypeError: this._init is not a function`.

The only workaround for now requires to make layout as basic Vue.js component, 
a minor inconvenience, but a problem nonetheless.

reinink commented:
This is interesting, because I was listening to the Full Stack Radio episode today with Evan You ([What's Coming in Vue.js 3.0](http://www.fullstackradio.com/129)), and it has me wondering what the implications are going to be for Inertia. I know Vue 3.0 will have much better TypeScript support, and I know they decided against the Class based component approach.

It has me wondering if we'll need a whole new adapter for Vue 3.0, or if we'll be able to maintain on package that supports both versions.

I'm going to do a little more research on this before I merge this in. I want to properly understand the implications first.

Thanks so much for putting this together though, and for your interest in Inertia! 🤟

TheDoctor0 commented:
I must say, wasn't expecting response so soon :)

Yes, it's true that Composition API in Vue.js 3.0 will provide much better TypeScript support.
However Class API proposal (that was rejected) is not the same as current class-style component syntax. It will still be [available](https://github.com/vuejs/rfcs/pull/17#issuecomment-494775371):
> Also, tro be clear: Vue 3 will continue to support the current object notation. We juist dropped the class API that we proposed as an additional API - and we dropped it for the reasons meantioned by Evan further up.

I created example project to reproduce issue: [inertia-vue-typescript-example](https://github.com/TheDoctor0/inertia-vue-typescript-example).
I added debug log to resolveComponent, so component object structure for resolved components can be easily analyzed. There are three branches:
1. master - layout component is ignored because it is not on _default_ level in returned object.
2. options - resolveComponent was changed to return correct object, but layout also has nested options property, so it throws `TypeError: this._init is not a function`.
It mirrors what was done in second commit in this PR but without modifying InertiaApp object.
With this minor change layout component that is not class-styled will work correctly.
3. fix - inertia-vue version was changed to presented in this PR, class-style layout component is fully supported.

Also, you can take a look at [inertia-shims.d.ts](https://github.com/TheDoctor0/inertia-vue-typescript-example/blob/master/resources/js/inertia-shims.d.ts). It provides better IDE support for Vue components with TypeScript. I was actually thinking yesterday about making the PR to add some information about it to documentation.

TheDoctor0 commented:
@reinink, any progress in your research? 
It seems that I am not the only one that would gladly see this change merged.

mattijsbliek commented:
@reinink I'm wondering what the plan is for this PR? I'm running into the situation that I have to make a choice between:

A. Use `Vue.extend` and not use a persistant layout.
B. Do not use `Vue.extend`. This fixes persistant layout issue, but Typescript no longer understands `this`, so I have to explicitly cast it everywhere.

TheDoctor0 commented:
I must admin that for me it was option C. Use Vue.extend with persistent layout, but from my fork.
It's far from perfect for production application, but in this case it was best solution.

GDanielRG commented:
Any plans on this? im stuck using class components and cant have layouts. Know of a quick fix while (hopefully) this PR gets merged?

TheDoctor0 commented:
For now, the only way is to fork this repository and apply those changes by yourself or use the fork I made for this PR:
```
"@inertiajs/inertia-vue": "https://github.com/TheDoctor0/inertia-vue#9f8999b",
```

@sebastiandedeyne, maybe you can review and merge it?

GDanielRG commented:
hmmm guess ill have to avoid using class component with Inertia where I am trying to use layouts for now... if any input or feedback helps speeding this up, I am willing to help however is needed. just started using Inertia and I am liking it so far, still learning things though.

reinink commented:
Hello, thanks so much for your interest in Inertia.js and for contributing back to it. However, in an effort to make the project easier to maintain, we've merged the four client-side libraries (`inertia`, `inertia-react`, `inertia-svelte` and `inertia-vue`) into a single [monorepo](https://github.com/inertiajs/inertia). If you feel this PR is still relevant, please resubmit it there. Sorry for the inconvenience. ❤️
